File: packages/dotnet/src/Webhooks.cs
using System.Globalization;
using System.Text;
using System.Text.Json;

namespace Catalisa.Biometrics;

/// <summary>Which check refused a delivery.</summary>
public enum WebhookFailure
{
    MissingHeaders,
    TimestampInvalid,
    TimestampOutOfTolerance,
    UnknownKeyId,
    UnsupportedSignatureVersion,
    InvalidSignature,
}

/// <summary>A delivery that failed verification. Answer 400 and do not process the body.</summary>
public sealed class WebhookVerificationException : Exception
{
    public WebhookFailure Failure { get; }

    public WebhookVerificationException(WebhookFailure failure, string message) : base(message)
    {
        Failure = failure;
    }
}

/// <summary>
/// Verifies Catalisa Webhooks Engine deliveries:
///
/// <code>
/// message   = x-webhook-id + "\n" + x-webhook-timestamp + "\n" + raw body
/// signature = RSA-SHA256 (PKCS#1 v1.5), base64, in x-webhook-signature as "v1=&lt;base64&gt;"
/// key       = the subscription's PUBLIC key, chosen by x-webhook-key-id
/// </code>
///
/// The signature is asymmetric: there is no shared secret to leak. The engine
/// does not refuse old deliveries — the receiver enforces the tolerance, or a
/// captured delivery could be replayed forever.
/// </summary>
public sealed class WebhookVerifier
{
    private readonly IReadOnlyDictionary<string, string> _publicKeys;
    private readonly TimeSpan _tolerance;
    private readonly Func<DateTimeOffset> _now;

    /// <param name="publicKeys">keyId to PEM. Keep a retired key here while deliveries signed with it may still arrive.</param>
    /// <param name="tolerance">Default five minutes. <see cref="TimeSpan.Zero"/> disables the check (not recommended).</param>
    /// <param name="now">Reference clock, for tests.</param>
    public WebhookVerifier(
        IReadOnlyDictionary<string, string> publicKeys,
        TimeSpan? tolerance = null,
        Func<DateTimeOffset>? now = null)
    {
        _publicKeys = publicKeys;
        _tolerance = tolerance ?? TimeSpan.FromMinutes(5);
        _now = now ?? (() => DateTimeOffset.UtcNow);
    }

    /// <summary>
    /// Verifies a delivery and returns the parsed event.
    ///
    /// Pass the RAW body — the bytes as they arrived. Decoding and re-encoding the
    /// JSON changes them, and the signature will not match.
    /// </summary>
    public WebhookEvent Verify(Func<string, string?> header, string rawBody)
    {
        var id = header("x-webhook-id");
        var timestamp = header("x-webhook-timestamp");
        var keyId = header("x-webhook-key-id");
        var signature = header("x-webhook-signature");

        if (string.IsNullOrEmpty(id) || string.IsNullOrEmpty(timestamp) || string.IsNullOrEmpty(keyId) || string.IsNullOrEmpty(signature))
        {
            throw Fail(WebhookFailure.MissingHeaders, "missing x-webhook-id, x-webhook-timestamp, x-webhook-key-id or x-webhook-signature");
        }
        if (!DateTimeOffset.TryParse(timestamp, CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal, out var sentAt))
        {
            throw Fail(WebhookFailure.TimestampInvalid, "x-webhook-timestamp is not an ISO 8601 date");
        }
        if (_tolerance > TimeSpan.Zero)
        {
            var drift = _now() - sentAt;
            if (drift > _tolerance || drift < -_tolerance)
            {
                throw Fail(WebhookFailure.TimestampOutOfTolerance, "delivery is outside the time tolerance (possible replay)");
            }
        }
        if (!signature!.StartsWith("v1=", StringComparison.Ordinal))
        {
            throw Fail(WebhookFailure.UnsupportedSignatureVersion, "unsupported signature version (expected v1=)");
        }
        if (!_publicKeys.TryGetValue(keyId!, out var pem))
        {
            throw Fail(WebhookFailure.UnknownKeyId, "no public key for this x-webhook-key-id");
        }

        byte[] raw;
        try
        {
            raw = Convert.FromBase64String(signature.Substring(3));
        }
        catch (FormatException)
        {
            throw Fail(WebhookFailure.InvalidSignature, "invalid signature");
        }

        var message = Encoding.UTF8.GetBytes($"{id}\n{timestamp}\n{rawBody}");
        bool valid;
        try
        {
            valid = Rsa.VerifySha256(pem, message, raw);
        }
        catch (Exception)
        {
            throw Fail(WebhookFailure.UnknownKeyId, "the public key for this keyId is not readable");
        }
        if (!valid)
        {
            throw Fail(WebhookFailure.InvalidSignature, "invalid signature");
        }

        try
        {
            return JsonSerializer.Deserialize<WebhookEvent>(rawBody)!;
        }
        catch (JsonException e)
        {
            throw Fail(WebhookFailure.InvalidSignature, "the signed body is not valid JSON: " + e.Message);
        }
    }

    private static WebhookVerificationException Fail(WebhookFailure failure, string message)
        => new(failure, message);
}

/// <summary>The outcome of checking an evidence signature offline.</summary>
public sealed record EvidenceVerification(bool Valid, string KeyId, bool UnknownKey, string? RetiredAt);

/// <summary>
/// OFFLINE verification of the signed evidence, without calling Catalisa:
///
/// <code>
/// message   = "{sessionId}|{attempt}|{bundleHash}|{signedAt}"
/// signature = Ed25519 (RFC 8032), base64
/// key       = GET /evidence-keys, by the signature's keyId
/// </code>
///
/// This is what lets a third party — an auditor, a court — confirm years later
/// that the bundle is the one Catalisa signed.
/// </summary>
public static class Evidence
{
    /// <summary>The exact string that was signed.</summary>
    public static string Message(string sessionId, int attempt, string bundleHash, string signedAt)
        => $"{sessionId}|{attempt}|{bundleHash}|{signedAt}";

    public static EvidenceVerification Verify(
        string sessionId,
        int attempt,
        string bundleHash,
        EvidenceSignature signature,
        IEnumerable<EvidenceKey> keys)
    {
        var key = keys.FirstOrDefault(k => k.KeyId == signature.KeyId);
        if (key is null)
        {
            return new EvidenceVerification(false, signature.KeyId, true, null);
        }
        if (signature.Alg != "Ed25519")
        {
            return new EvidenceVerification(false, signature.KeyId, false, key.RetiredAt);
        }

        byte[] raw;
        byte[] publicKey;
        try
        {
            raw = Convert.FromBase64String(signature.Value);
            publicKey = RawKeyFromPem(key.PublicKeyPem);
        }
        catch (FormatException)
        {
            return new EvidenceVerification(false, signature.KeyId, false, key.RetiredAt);
        }

        var message = Encoding.UTF8.GetBytes(Message(sessionId, attempt, bundleHash, signature.SignedAt));
        var valid = Ed25519.Verify(publicKey, message, raw);

        return new EvidenceVerification(valid, signature.KeyId, false, key.RetiredAt);
    }

    /// <summary>
    /// The 32 raw bytes of an Ed25519 key inside a PEM. The DER wrapper
    /// (SubjectPublicKeyInfo) is a fixed 12-byte header followed by the key, which
    /// is why taking the tail is enough — and why a key of another kind, being a
    /// different length, is refused here.
    /// </summary>
    private static byte[] RawKeyFromPem(string pem)
    {
        var body = pem
            .Replace("-----BEGIN PUBLIC KEY-----", "")
            .Replace("-----END PUBLIC KEY-----", "")
            .Replace("\r", "")
            .Replace("\n", "")
            .Trim();
        var der = Convert.FromBase64String(body);
        if (der.Length != 44)
        {
            throw new FormatException("not an Ed25519 public key");
        }

        return der.Skip(12).ToArray();
    }
}

File: packages/go/webhooks.go
package biometrics

import (
	"crypto"
	"crypto/ed25519"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"errors"
	"fmt"
	"net/http"
	"strings"
	"time"
)

// Webhook deliveries are signed by the Catalisa Webhooks Engine:
//
//	message   = x-webhook-id + "\n" + x-webhook-timestamp + "\n" + raw body
//	signature = RSA-SHA256 (PKCS#1 v1.5), base64, in x-webhook-signature as "v1=<base64>"
//	key       = the subscription's PUBLIC key, chosen by x-webhook-key-id
//
// The signature is asymmetric: there is no shared secret to leak. The server
// does not reject old deliveries, so the receiver enforces the time tolerance.

// WebhookFailure says why a delivery was refused.
type WebhookFailure string

const (
	FailureMissingHeaders       WebhookFailure = "MISSING_HEADERS"
	FailureTimestampInvalid     WebhookFailure = "TIMESTAMP_INVALID"
	FailureTimestampOutOfRange  WebhookFailure = "TIMESTAMP_OUT_OF_TOLERANCE"
	FailureUnknownKeyID         WebhookFailure = "UNKNOWN_KEY_ID"
	FailureUnsupportedSignature WebhookFailure = "UNSUPPORTED_SIGNATURE_VERSION"
	FailureInvalidSignature     WebhookFailure = "INVALID_SIGNATURE"
)

// ErrWebhookVerification wraps every refusal, so errors.Is can catch them all.
var ErrWebhookVerification = errors.New("webhook verification failed")

// WebhookError says which check failed. Answer 400 and do not process the body.
type WebhookError struct {
	Failure WebhookFailure
	Message string
}

func (e *WebhookError) Error() string {
	return fmt.Sprintf("biometrics webhook: %s: %s", e.Failure, e.Message)
}
func (e *WebhookError) Unwrap() error { return ErrWebhookVerification }

var failureMessages = map[WebhookFailure]string{
	FailureMissingHeaders:       "missing x-webhook-id, x-webhook-timestamp, x-webhook-key-id or x-webhook-signature",
	FailureTimestampInvalid:     "x-webhook-timestamp is not an ISO 8601 date",
	FailureTimestampOutOfRange:  "delivery is outside the time tolerance (possible replay)",
	FailureUnknownKeyID:         "no public key for this x-webhook-key-id",
	FailureUnsupportedSignature: "unsupported signature version (expected v1=)",
	FailureInvalidSignature:     "invalid signature",
}

func webhookErr(f WebhookFailure) *WebhookError {
	return &WebhookError{Failure: f, Message: failureMessages[f]}
}

// WebhookVerifier verifies deliveries against the subscription's public keys,
// keyed by their keyId (`GET /webhooks-engine/api/v1/subscriptions/{id}/keys`).
// A retired key stays in the map so deliveries in flight still verify.
type WebhookVerifier struct {
	// Keys maps keyId to the PEM public key.
	Keys map[string]string
	// Tolerance for the delivery timestamp. Zero means 5 minutes; use a negative
	// value to disable the check (not recommended — it allows replay).
	Tolerance time.Duration
	// now replaces the clock in tests.
	now func() time.Time
}

// NewWebhookVerifier builds a verifier with the default 5-minute tolerance.
func NewWebhookVerifier(keys map[string]string) *WebhookVerifier {
	return &WebhookVerifier{Keys: keys}
}

// Verify checks the signature over the RAW body — the bytes as they arrived. Do
// not decode and re-encode the JSON before verifying: any reordering or
// reformatting changes the bytes and the signature will not match.
//
// On success it returns the parsed event. Use event.ID for idempotency: a
// delivery can arrive more than once.
func (v *WebhookVerifier) Verify(header http.Header, body []byte) (*WebhookEvent, error) {
	id := header.Get("x-webhook-id")
	timestamp := header.Get("x-webhook-timestamp")
	keyID := header.Get("x-webhook-key-id")
	signature := header.Get("x-webhook-signature")
	if id == "" || timestamp == "" || keyID == "" || signature == "" {
		return nil, webhookErr(FailureMissingHeaders)
	}

	sentAt, err := time.Parse(time.RFC3339Nano, timestamp)
	if err != nil {
		return nil, webhookErr(FailureTimestampInvalid)
	}
	tolerance := v.Tolerance
	if tolerance == 0 {
		tolerance = 5 * time.Minute
	}
	if tolerance > 0 {
		now := time.Now
		if v.now != nil {
			now = v.now
		}
		if drift := now().Sub(sentAt); drift > tolerance || drift < -tolerance {
			return nil, webhookErr(FailureTimestampOutOfRange)
		}
	}

	if !strings.HasPrefix(signature, "v1=") {
		return nil, webhookErr(FailureUnsupportedSignature)
	}
	pemKey, ok := v.Keys[keyID]
	if !ok {
		return nil, webhookErr(FailureUnknownKeyID)
	}
	publicKey, err := parsePublicKey(pemKey)
	if err != nil {
		return nil, webhookErr(FailureUnknownKeyID)
	}
	raw, err := base64.StdEncoding.DecodeString(strings.TrimPrefix(signature, "v1="))
	if err != nil {
		return nil, webhookErr(FailureInvalidSignature)
	}

	message := make([]byte, 0, len(id)+len(timestamp)+len(body)+2)
	message = append(message, id...)
	message = append(message, '\n')
	message = append(message, timestamp...)
	message = append(message, '\n')
	message = append(message, body...)
	digest := sha256.Sum256(message)

	rsaKey, ok := publicKey.(*rsa.PublicKey)
	if !ok {
		return nil, webhookErr(FailureInvalidSignature)
	}
	if err := rsa.VerifyPKCS1v15(rsaKey, crypto.SHA256, digest[:], raw); err != nil {
		return nil, webhookErr(FailureInvalidSignature)
	}

	var event WebhookEvent
	if err := json.Unmarshal(body, &event); err != nil {
		return nil, &WebhookError{Failure: FailureInvalidSignature, Message: "signed body is not valid JSON: " + err.Error()}
	}
	return &event, nil
}

func parsePublicKey(pemKey string) (any, error) {
	block, _ := pem.Decode([]byte(pemKey))
	if block == nil {
		return nil, errors.New("invalid PEM")
	}
	return x509.ParsePKIXPublicKey(block.Bytes)
}

// EvidenceVerification is the outcome of checking a signature offline.
type EvidenceVerification struct {
	Valid bool
	KeyID string
	// UnknownKey is true when no published key has this keyId.
	UnknownKey bool
	// RetiredAt is set when the signing key has since been retired. A signature
	// made before the retirement is still valid — this is for your audit trail.
	RetiredAt *string
}

// VerifyEvidence checks the evidence signature WITHOUT calling Catalisa. The
// message is `sessionID|attempt|bundleHash|signedAt`, signed with Ed25519.
// Attempt comes from the session envelope.
//
// This is what lets a third party — an auditor, a court — confirm that the
// decision bundle is the one Catalisa signed, years later, offline.
func VerifyEvidence(sessionID string, attempt int, bundleHash string, signature EvidenceSignature, keys []EvidenceKey) EvidenceVerification {
	out := EvidenceVerification{KeyID: signature.KeyID}
	var key *EvidenceKey
	for i := range keys {
		if keys[i].KeyID == signature.KeyID {
			key = &keys[i]
			break
		}
	}
	if key == nil {
		out.UnknownKey = true
		return out
	}
	out.RetiredAt = key.RetiredAt
	if signature.Alg != "Ed25519" {
		return out
	}
	parsed, err := parsePublicKey(key.PublicKeyPEM)
	if err != nil {
		return out
	}
	edKey, ok := parsed.(ed25519.PublicKey)
	if !ok {
		return out
	}
	raw, err := base64.StdEncoding.DecodeString(signature.Value)
	if err != nil {
		return out
	}
	message := fmt.Sprintf("%s|%d|%s|%s", sessionID, attempt, bundleHash, signature.SignedAt)
	out.Valid = ed25519.Verify(edKey, []byte(message), raw)
	return out
}
